### ISSUE TITLE
TWE Design System: add accessibility testing to Microsite

### ISSUE THREAD

**Issue by joshfishman** [hackforla/internship#554]
### Overview
We need to setup testing for the Design System microsite so we can automate accessibility testing and ensure that the design system meets accessibility standards

### Action Items
- [x]  Research best practices for including a testing library in your project
- [x] Research the best testing library for our needs, a few ideas below
- [x] Select a library and include in the project
- [x] process the list of errors with the dev team lead to figure out if new issues need to be made
- [x] Create documentation on how to test each UI commit, 
- [x] Add documentation to sub-page of software engineer team wiki

### Resources/Instructions
- [Google Sheet with Pa11y output](https://docs.google.com/spreadsheets/d/11vPqp_xwGGJi1LcDjwOK8b_z_QWUC4H-c7VsjgEHJpg/edit#gid=1748429059)
- [Team Wiki](https://github.com/hackforla/internship/wiki/Software-Engineer)
- [selenium](https://www.selenium.dev/)
- [Google Lighthouse node](https://developer.chrome.com/docs/lighthouse/overview/#programmatic)
- [Microsoft](https://accessibilityinsights.io/)
- [Pa11y](https://pa11y.org/)
- [Tanaguru](https://my.tanaguru.com/#/external-module/Home)

**Comment by aconstas** [comment]
4/13/24 
Is there a specific accessibility standard that we are trying to comply with? [WCAG2A, WCAG2AA, or WCAG2AAA](https://www.w3.org/WAI/standards-guidelines/)

What type of messages do we want from pa11y? [notice? warning? error?](https://bitsofco.de/pa11y/#only-display-critical-errors)

**Comment by aconstas** [comment]
![image (1)](https://github.com/hackforla/internship/assets/82304192/153c612e-5861-4642-95d9-e7d98115fb9a)

Pa11y identified several issues with how mkdocs generates a page from markdown files. One notable error that was logged in my integrated terminal concerns an anchor (<a>) element in the HTML. The error message from pa11y stated: **"Anchor element found with a valid href attribute, but no link content has been supplied."** 

It's unclear how mkdocs handles these elements internally, but pa11y flagged this as a problem because it can affect website accessibility.

Pa11y cited: [WCAG2AA.Principle1.Guideline1_3.1_3_1.F68](https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/F68#F68-examples)

### Is this an error we are concerned about? Or should I have Pa11y ignore these types of errors?

**Comment by aconstas** [comment]
- Commits will no longer be blocked if they do not pass the Pa11y accessibility test
- Looking into the best way to organize and gather Pa11y error messages (CSV, some type of log, etc.)

**Comment by ExperimentsInHonesty** [comment]
- First step for documenting the notice/warnings/error: a comprehensive list of every one across all the current pages in the repo (static)
- Evaluate results, determining which of the list we are going to address and how
- once eval and plan is created, we will know what kind of ongoing notifications that we want
- tune the notifications
- decide what documentation we need about errors we intend to address

**Comment by ExperimentsInHonesty** [comment]
On the spreadsheet, make a tab for all the pages, and add a column to indicate which page, so that we can sort by type of error (in code column).  This will help with the evaluation.

**Comment by aconstas** [comment]
### Research
I looked for Visual Studio Code accessibility plug-ins and found none that would meet our needs. The plug-ins I did find were linters that would not export error/warning messages (e.g., [Axes](https://marketplace.visualstudio.com/items?itemName=deque-systems.vscode-axe-linter))

### Moving Forward
- Continue with the setup for Pa11y
- Utilize Google Sheets API to store Pa11y output

### Questions
- How should Pa11y output be handled with Google Sheets? 
Suppose there's a pre-existing Google sheet full of accessibility errors, and a dev modifies an existing microsite page to handle some of those errors. Is it possible to overwrite data for that specific microsite page within the sheet? Or should Pa11y test the entire microsite, and override the pre-existing Google sheet completely?

**Comment by ExperimentsInHonesty** [comment]
Next step is to the create the spreadsheet which is all the errors on one sheet.

**Comment by aconstas** [comment]
Created a single [Google Sheet](https://docs.google.com/spreadsheets/d/11vPqp_xwGGJi1LcDjwOK8b_z_QWUC4H-c7VsjgEHJpg/edit#gid=1748429059) with errors from the entire microsite.

### Findings 
| Code | Count | Message |
| ------------- | ------------- | ------------- |
| WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.A.NoContent  | 1,337  | Anchor element found with a valid href attribute, but no link content has been supplied.  |
| WCAG2AA.Principle3.Guideline3_2.3_2_2.H32.2  | 8  | This form does not contain a submit button, which creates issues for those who cannot submit the form using the keyboard. |
|WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.InputCheckbox.Name |6 |This checkboxinput element does not have a name available to an accessibility API. Valid names are: label element, title undefined, aria-label undefined, aria-labelledby undefined. |
|WCAG2AA.Principle1.Guideline1_3.1_3_1.F68 | 28 |This form field should be labelled in some way... |
|WCAG2AA.Principle1.Guideline1_4.1_4_3.G145.Fail | 9| This element has insufficient contrast at this conformance level. Expected a contrast ratio of at least 3:1... |
|WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.InputText.Name |9 |This textinput element does not have a name available to an accessibility API. Valid names are: label element, title undefined, aria-label undefined ... |
|WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.InputRadio.Name | 10 | This radioinput element does not have a name available to an accessibility API. Valid names are: label element, title undefined, aria-label undefined ... |
|WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.Textarea.Name | 3| This textarea element does not have a name available to an accessibility API. Valid names are: label element, title undefined, aria-label undefined, aria-labelledby undefined. |

**Comment by ExperimentsInHonesty** [comment]
@aconstas 
Please provide update
1. Progress: "What is the current status of your project? What have you completed and what is left to do?"
2. Blockers: "Difficulties or errors encountered."
3. Availability: "How much time will you have this week to work on this issue?"
4. ETA: "When do you expect this issue to be completed?"
5. Pictures or links* (if necessary): "Add any pictures or links that will help illustrate what you are working on."

* remember to add links to the top of the issue if they are going to be needed again.

**Comment by aconstas** [comment]
1. **Progress**: "What is the current status of your project? What have you completed and what is left to do?"
       Completed the collection of all errors from microsite. We need to discuss what errors we are focused on and what errors we can ignore. Then we need to discus the best way to implement the logging of the errors to Google Sheets.
   2. **Blockers**: None
   3. **Availability**: all week
   4. **ETA**: 5/23 or sooner depending on when we can discuss
   5. **Pictures or links**:  [Google Sheet](https://docs.google.com/spreadsheets/d/11vPqp_xwGGJi1LcDjwOK8b_z_QWUC4H-c7VsjgEHJpg/edit#gid=1748429059)

**Comment by aconstas** [comment]
## Progress:
Josh and I met today and planned to implement `aria-label`s for the majority of issues. Here is the plan thus far:
![image](https://github.com/hackforla/internship/assets/82304192/0f738008-f968-4ed1-8715-a1bd5edfd1c6)

I have started exploring how to rectify the many **WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.A.NoContent** errors.

I added a small JS script in the extra extra.js file that selects all elements that are `<a>` tags and children of `<span>` tags with an href containing "__codelineno". MkDocs utilizes these anchor tags with hrefs including "codelineno" to construct codeblocks. 

After DOMContentLoaded, the script selects all the codeline anchors and removes that href attribute. I have confirmed that this does not affect any functionality (e.g., copy codeblock button).

### From the main page of the site (/internship-website-design-system) : 
**Before:**
![beforeJSHack](https://github.com/hackforla/internship/assets/82304192/39457cdb-f24b-4f36-91b1-2f24889019a8)

**After:**
![afterJSHack](https://github.com/hackforla/internship/assets/82304192/441fbc9d-f766-4080-9fb3-6bde525e812a)

## Blocker
The JavaScript hack works with Pa11y because Pa11y starts a headless browser, navigates to the URL, and runs the accessibility test. However, when a user navigates through the site via navigation links, the same JavaScript hack is not applied to the new page. This might be fixed by using an event listener on `window.location`. I will look into this.

**Comment by aconstas** [comment]
I have resolved most of the accessibility errors and have 30 left.
|  Code |  Count |
|  ----------------------------- |   ----------------------------- |
| WCAG2AA.Principle1.Guideline1_4.1_4_3.G145.Fail | 9 |
| WCAG2AA.Principle1.Guideline1_4.1_4_3.G18.Fail | 1 |
| WCAG2AA.Principle3.Guideline3_2.3_2_2.H32.2 | 8 |
| WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.InputCheckbox.Name | 6 |
| WCAG2AA.Principle1.Guideline1_3.1_3_1.F68 | 6 |

**G145** and **G18** are insufficient contrast design issues.

## Blocker
**H91** and **F68** come from two checkboxes that are included in the DOM on every page of the site. It appears that these checkboxes are tied to toggling the side menu and the search box. In all instances of the two checkboxes, they have `display: none`. I'm not sure how it functions with MkDocs or if it's generated but I'm going to look into it. Either way, if we put an `aria-label` on them it will resolve the accessibility issue. Curious though about where these checkboxes come from and why they're set up the way they are.

**H32.2** comes from the visible search box at the top of the page. I think adding an `aria-label` would suffice for that.

**Comment by aconstas** [comment]
Getting ready to submit a PR for this feat.

- Pre-Commit hook runs pa11y accessibility test and outputs results as a CSV file in the `/output` folder
- Pre-Commit hook will **not** block commits
- Devs can bypass the pre-commit hook if need be with `git commit -m 'message -n`
- Devs can run the accessibility check with `npm run pa11y` if they want to monitor accessibility errors as they work before making any commits
- View CSV result file within your editor with the [Excel Viewer by GrapeCity](https://marketplace.visualstudio.com/items?itemName=GrapeCity.gc-excelviewer) plugin. ( Supports organizing and filtering errors)

**Comment by aconstas** [comment]
Resolved the majority of the accessibility errors with [PR #7](https://github.com/hackforla/internship-website-design-system/pull/7)

Code: WCAG2AA.Principle3.Guideline3_2.3_2_2.H32.2
Message: "This form does not contain a submit button, which creates issues for those who cannot submit the form using the keyboard. Submit buttons are INPUT elements with type attribute "submit" or "image", or BUTTON elements with type "submit" or omitted/invalid."

This is the search input that is generated by MkDocs. The `<form>` the search input is contained in does not have a submit button. I was initially thinking that adding a hidden button would solve the issue but from thinking about the problem from the standpoint of someone with accessibility issues, that would not be helpful.


This has been a known issue since 2017 and doesn't look like it will ever be implemented in the theme: https://github.com/squidfunk/mkdocs-material/issues/400#issuecomment-316968253

### Should we override the material theme with a custom header section and include a button with the search box?

**Comment by aconstas** [comment]
Fixed the search input accessibility issue by overriding the partial (search.html) from the material theme.
- Created /docs/overrides
- search.html: Created a hidden button with `aria-label: "hidden"` and `display: none`

0 accessibility issues with the site as of now.
